internal/configs/provider_validation.go

```diff
@@ -95,6 +142,44 @@ func validateProviderConfigs(parentCall *ModuleCall, cfg *Config, noProviderConf
 		}
 	}
 
+	checkImpliedProviderNames := func(resourceConfigs map[string]*Resource) {
+		// Now that we have all the provider configs and requirements validated,
+		// check for any resources which use an implied localname which doesn't
+		// match that of required_providers
+		for _, r := range resourceConfigs {
+			// We're looking for resources with no specific provider reference
+			if r.ProviderConfigRef != nil {
+				continue
+			}
```

Comment:
Aren't we _also_ looking for resources that explicitly specify a provider local name that wasn't declared?

```hcl
  provider = aws.foo
```

I believe (though I haven't tested) that if we encounter the above without there being an `aws` in `required_providers` then we treat it as an implied requirement for `hashicorp/aws`, just as we would if it were not set at all but the type name begins with "aws".

[`Resource.ProviderConfigAddr`](https://github.com/hashicorp/terraform/blob/9db277af427d18af078cd07915733d04e14daf71/internal/configs/resource.go#L72-L90) seems to be the definition of which local name a particular resource is associated with, and so maybe we can use that here and then take the `LocalName` field of the result as the local name to use below.